agent/grpc/resolver/resolver.go

```diff
@@ -0,0 +1,287 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/metadata"
+	"github.com/hashicorp/consul/agent/router"
+	"google.golang.org/grpc/resolver"
+)
+
+var registerLock sync.Mutex
+
+// RegisterWithGRPC registers the ServerResolverBuilder as a grpc/resolver.
+// This function exists to synchronize registrations with a lock.
+// grpc/resolver.Register expects all registration to happen at init and does
+// not allow for concurrent registration. This function exists to support
+// parallel testing.
+func RegisterWithGRPC(b *ServerResolverBuilder) {
+	registerLock.Lock()
+	defer registerLock.Unlock()
+	resolver.Register(b)
+}
+
+// Nodes provides a count of the number of nodes in the cluster. It is very
+// likely implemented by serf to return the number of LAN members.
+type Nodes interface {
+	NumNodes() int
+}
+
+// ServerResolverBuilder tracks the current server list and keeps any
+// ServerResolvers updated when changes occur.
+type ServerResolverBuilder struct {
+	// datacenter of the local agent.
+	datacenter string
+	// scheme used to query the server. Defaults to consul. Used to support
+	// parallel testing because gRPC registers resolvers globally.
+	scheme string
+	// servers is an index of Servers by Server.ID
+	// TODO: this assumes that server.ID is unique across all datacenter. Is that
+	// a safe assumption?
+	servers map[string]*metadata.Server
```

Comment:
Calling out this TODO. I ran into this problem while writing tests. I think the ID comes from  serf `Tags["id"]`, but I'm not sure where that gets set. Is this the NodeID? Is this assumption about the nodeID being globally unique across all DC a safe one?

Reply:
One way to resolve this would be to stop tracking servers separately in `ResolverBuilder` and ask for the list from Router (ex: 68af47b263dfd921074a3c8fb485331254f3c4ef).  I think something like that might be nice, because we store the list of servers in fewer places.

Unfortunately it introduces a circular dependency, which would require more complicate notifications of changes, so maybe not worth it right now.

Reply:
Ok, everything I have tried so far ends up being a lot more complicated, so I think it'll be good to keep it this way for now, and I will document the assumption. With some larger refactoring to Router it might be possible to share more of the code, but I'm not sure.